dev/com.ibm.jbatch.container/src/com/ibm/jbatch/container/services/impl/JPAPersistenceManagerImpl.java

```diff
@@ -2685,34 +2698,169 @@ private int getJobInstanceTableVersion(PersistenceServiceUnit psu) throws Except
             while (cause != null) {
                 if (cause instanceof SQLSyntaxErrorException &&
                     cause.getMessage() != null &&
-                    cause.getMessage().contains("UPDATETIME")) {
+                    cause.getMessage().contains("GROUPNAMES")) {
                     // The column isn't there.
-                    logger.fine("The UPDATETIME column does not exist, job instance table version = 1");
-                    instanceVersion = 1;
-                    return instanceVersion;
+                    try {
+                        Integer exec = new TranRequest<Integer>(em) {
+                            @Override
+                            public Integer call() throws Exception {
+                                // Verify that UPDATETIME column exists by running a query against it.
+                                String queryString = "SELECT COUNT(x.lastUpdatedTime) FROM JobInstanceEntityV2 x";
+                                TypedQuery<Long> query = em.createQuery(queryString, Long.class);
+                                query.getSingleResult();
+                                logger.fine("The UPDATETIME column exists, job instance table version = 2");
+                                instanceVersion = 2;
+                                return instanceVersion;
+                            }
+                        }.runInNewOrExistingGlobalTran();
+
+                        return exec;
+                    } catch (javax.persistence.PersistenceException e2) {
+                        cause = e2.getCause();
+                        while (cause != null) {
+                            if (cause instanceof SQLSyntaxErrorException &&
+                                cause.getMessage() != null &&
+                                cause.getMessage().contains("UPDATETIME")) {
+                                // The column isn't there.
+                                logger.fine("The UPDATETIME column does not exist, job instance table version = 1");
+                                instanceVersion = 1;
+                                return instanceVersion;
+                            }
+                            cause = cause.getCause();
+                        }
+                    }
                 }
-                cause = cause.getCause();
             }
             logger.fine("Unexpected exception while checking job instance table version, re-throwing");
-            throw e;
+            throw e3;
         } finally {
             em.close();
         }
     }
+
+    @Override
+    public List<String> getGroupNamesForJobID(long jobInstanceID) throws NoSuchJobInstanceException {
+        EntityManager em = getPsu().createEntityManager();
+        try {
+            TypedQuery<JobInstanceEntityV3> query = em.createNamedQuery(JobInstanceEntityV3.GET_ALL_GROUPNAMES_BY_JOBINSTANCE_QUERY,
+                                                                        JobInstanceEntityV3.class);
```

Comment:
I'm confused...why is the result type a JobInstanceEntityV3 instead of a String?  (since it's group name you're after), like the return type of this method itself.

Reply:
Wouldn't just a regular `find` by instanceId work, followed by returning the groupNames?   Doesn't JPA and the DB just load this for you already?    Or am I missing something?

Reply:
We can get rid of this query:
TypedQuery<JobInstanceEntityV3> query = em.createNamedQuery(JobInstanceEntityV3.GET_ALL_GROUPNAMES_BY_JOBINSTANCE_QUERY,
                                                                        JobInstanceEntityV3.class);

and replace it with a call to the generic find

Reply:
method removed and all references to it switched over.